File: leetcode/atoi.py
import sys

def charToDigit(c):
	return ord(c) - ord('0')

def atoi(str):
	sign = True
	if (str[0] == '-'):
		sign = False
		str = str[1:]
	result = 0
	overflow = False
	for i in xrange(len(str)):
		digit = charToDigit(str[i])
		if result >= (sys.maxint - digit)/10:
			result = long(result)
		result = result * 10 + digit
	if not sign:
		result = -result
	return result

assert(atoi("-12345678901") == -12345678901)
assert(atoi("0") == 0)
assert(atoi("99999999999999999999999999999999999") == 99999999999999999999999999999999999L)

File: leetcode/atoi.cpp
#include <cstdlib>
#include <climits>
#include <iostream>
#include <cctype>
using namespace std;

class Solution {
public:
  int
  atoi(const char *str) {
    bool sign = true;
    while (*str == ' ') ++str;
    if (*str == '-') {
      sign = false;
      ++str;
    } else if (*str == '+') ++str;

    int result = 0;
    bool overflow = false;
    for (; *str != '\0'; ++str) {
      if (!isdigit(*str)) break;
      int n = *str - '0';
      if ((INT_MAX - n)/10 < result) {
        overflow = true;
        break;
      }
      result = 10*result + n;
    }
    if (overflow) return sign ? INT_MAX : INT_MIN;
    return sign ? result : -result;
  }
};

int main(int argc, char** argv) {
  Solution s;
  //cout << s.atoi(argv[1]) << endl;
  cout << INT_MAX << endl;
  cout << INT_MIN << endl;
  return EXIT_SUCCESS;
}
